default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1602 @@
+        except Exception:
+            self.part_by_partname.clear()
+            traceback.print_exc()
+
+        corrupted = []
+        for designname in self.design_id_by_name:
+            try:
+                cached_name = fo.getShipDesign(self.design_id_by_name[designname]).name(False)
+                if cached_name != designname:
+                    print "WARNING: ShipID cache corrupted."
+                    print "Expected: %s, got: %s. Repairing cache." % (designname, cached_name)
+                    design_id = next(iter([shipDesignID for shipDesignID in fo.getEmpire().allShipDesigns
+                                          if designname == fo.getShipDesign(shipDesignID).name(False)]), None)
+                    if design_id is not None:
+                        self.design_id_by_name[designname] = design_id
+                    else:
+                        corrupted.append(designname)
+            except AttributeError:
+                print "WARNING: ShipID cache corrupted. Could not get cached shipdesign. Repairing Cache."
+                print traceback.format_exc()  # do not print to stderr as this is an "expected" exception.
+                design_id = next(iter([shipDesignID for shipDesignID in fo.getEmpire().allShipDesigns
+                                      if designname == fo.getShipDesign(shipDesignID).name(False)]), None)
+                if design_id is not None:
+                    self.design_id_by_name[designname] = design_id
+                else:
+                    corrupted.append(designname)
+        for corrupted_entry in corrupted:
+            del self.design_id_by_name[corrupted_entry]
+
+    def _update_buildable_items_this_turn(self, verbose=False):
+        """Calculate which parts and hulls can be built on each planet this turn.
+
+        :param verbose: bool - toggles detailed debugging output.
+        """
+        # The AI currently has no way of checking building requirements of individual parts and hulls directly.
+        # It can only check if we can build a design. Therefore, we use specific testdesigns to check if we can
+        # build a hull or part.
+        # The building requirements are constant so calculate this only once at the beginning of each turn.
+        #
+        # Code structure:
+        #   1. Update hull test designs
+        #   2. Get a list of buildable ship hulls for each planet
+        #   3. Update ship part test designs
+        #   4. Cache the list of buildable ship parts for each planet
+        #
+        self.hulls_for_planets.clear()
+        self.parts_for_planets.clear()
+        get_shipdesign = fo.getShipDesign
+        get_hulltype = fo.getHullType
+        empire = fo.getEmpire()
+        empire_id = empire.empireID
+        universe = fo.getUniverse()
+        planets_with_shipyards = _get_planets_with_shipyard()
+        available_hulls = list(empire.availableShipHulls)   # copy so we can sort it locally
```

Comment:
I could split it. However, the expected behaviour is that the ```TESTDESIGN_PREFERED_HULL``` has no tech requirements. Thus not being in ```available_hulls``` is an exception and is treated as such. I'd even make the point it not being found is an implementation error and the AI needs to be updated as such.

I move the prefered testhull to the front in order to minimize the number of lookups in the list.
```hulls_for_planet``` receives the same ordering. Later on, to test the buildability of parts I need to find a hull which can mount the part (i.e. has the corresponding slot). I do this by looking up each available hull for the planet. Now if "bad" hulls are in the front of the list (let's say multiple stuff that has only external oder only internal slots), I need to iterate over the list some times for each shipyard. This should most certainly be slower than a single swap operation inside the list.

Arguably, this is probably a very minor effect in the current code. It is mostly a legacy artifact from an old and quite unoptimized algorithm which I simply did not care enough about to take out.